File: backend/app/domain/reminders.py
"""A-7 期限リマインダーの bucket 判定 + 基準日 helper (ADR-00045).

純粋なドメインロジックのみを持つ (DB / FastAPI 非依存)。reminder endpoint と date_context
endpoint がこの module の `compute_reminder_bucket` / `_today_jst` / 定数を共有することで、
bucket 判定と "today" の権威を 1 箇所に集約する (plan-review R1 F-002 / R2 F-001)。

不変条件:
- `due_date` は ADR-00034 の暦日 (`date`、timezone なし)。"today" も固定 timezone (Asia/Tokyo) の暦日。
- bucket 判定は `compute_reminder_bucket(due_date, reference_date, threshold_days)` が正本。
  SQL prefilter (`due_date <= reference_date + threshold_days`) に依存せず、window 外を `None`
  で判定できる (R2 F-001)。
- frontend `dueDateBucket` (`frontend/lib/domain/due-date.ts`) は同一 signature / 同一 4 値を返す。
"""

from __future__ import annotations

from datetime import date, datetime
from typing import Literal
from zoneinfo import ZoneInfo

# upcoming (期限近接) とみなす窓 (日数)。reference_date + この日数 までを upcoming とする。
# 変更は ADR-00045 の更新を要する (backend 固定、caller 入力不可)。
REMINDER_UPCOMING_WINDOW_DAYS = 7

# bucket ごとに独立して返す items の上限 (plan-review R1 F-001)。全 bucket 横断の単一 LIMIT は
# 採らない (overdue が due_today / upcoming の表示枠を枯渇させる silent truncation を防ぐ)。
REMINDER_BUCKET_LIST_LIMIT = 50

# 期限の暦日を解釈する timezone。due_date は時刻概念のない暦日のため、"today" も JST 暦日で固定する。
_JST = ZoneInfo("Asia/Tokyo")

ReminderBucketName = Literal["overdue", "due_today", "upcoming"]


def today_jst(now: datetime | None = None) -> date:
    """Asia/Tokyo の暦日 ("today") を返す。

    reminder endpoint と date_context endpoint が共有する "today" の唯一権威 (R1 F-002 / R2 F-002)。
    ``now`` は test 用の注入点 (省略時は現在時刻)。JST 深夜 0 時境界を決定的に検証できる。
    """
    instant = now if now is not None else datetime.now(tz=_JST)
    return instant.astimezone(_JST).date()


def compute_reminder_bucket(
    due_date: date,
    reference_date: date,
    threshold_days: int = REMINDER_UPCOMING_WINDOW_DAYS,
) -> ReminderBucketName | None:
    """due_date を基準日 + 閾値から bucket 分類する (bucket 判定の正本、R2 F-001)。

    - ``due_date < reference_date`` -> ``"overdue"`` (下限なし、古い超過も超過)
    - ``due_date == reference_date`` -> ``"due_today"``
    - ``reference_date < due_date <= reference_date + threshold_days`` -> ``"upcoming"``
    - ``due_date > reference_date + threshold_days`` -> ``None`` (window 外、reminder 対象外)

    SQL の prefilter と同一 ``threshold_days`` を渡すこと。frontend ``dueDateBucket`` と同じ判定。
    """
    days_until = (due_date - reference_date).days
    if days_until < 0:
        return "overdue"
    if days_until == 0:
        return "due_today"
    if days_until <= threshold_days:
        return "upcoming"
    return None


__all__ = [
    "REMINDER_BUCKET_LIST_LIMIT",
    "REMINDER_UPCOMING_WINDOW_DAYS",
    "ReminderBucketName",
    "compute_reminder_bucket",
    "today_jst",
]

File: frontend/lib/api/reminders.ts
import { z } from "zod";

import { fetchBackendJson } from "@/lib/api/client";
import { isValidYmd } from "@/lib/domain/due-date";

// A-7 (ADR-00045): 期限リマインダー集約 (read-only、on-read 派生) + 一覧用 date_context。
// response 全 field を Zod で必須検証 (data 完全性: 不完全を完全と見せない)。malformed / auth 失効 /
// schema drift は parse で throw し、呼出側が degraded (ok/error) に倒す (fail-closed)。

// 実在する暦日 (YYYY-MM-DD) のみ許容 (adversarial R1 F-001)。非実在日 (2026-13-40 等) を取得失敗に
// 倒し、基準日 authority が壊れた状態で赤/橙を誤表示しない。
const ymdString = z.string().refine(isValidYmd, { message: "invalid calendar date (YYYY-MM-DD)" });
// reminder window は非負かつ妥当な上限内 (壊れた threshold で window を誤拡大しない)。
const thresholdDaysSchema = z.number().int().min(0).max(366);

const ReminderItemSchema = z.object({
  ticket_id: z.string(),
  project_id: z.string(),
  slug: z.string(),
  title: z.string(),
  status: z.string(),
  priority: z.string().nullable(),
  due_date: ymdString,
  days_until: z.number().int()
});

export type ReminderItem = z.infer<typeof ReminderItemSchema>;

// bucket 別 (plan-review R1 F-001): count は SQL COUNT (正確)、items は bucket ごとに独立 cap、
// truncated は count > items.length。overdue が due_today / upcoming の items を枯渇させない。
const ReminderBucketSchema = z.object({
  count: z.number().int(),
  truncated: z.boolean(),
  items: z.array(ReminderItemSchema)
});

export type ReminderBucket = z.infer<typeof ReminderBucketSchema>;

export const ReminderSummarySchema = z.object({
  reference_date: ymdString,
  threshold_days: thresholdDaysSchema,
  overdue: ReminderBucketSchema,
  due_today: ReminderBucketSchema,
  upcoming: ReminderBucketSchema
});

export type ReminderSummary = z.infer<typeof ReminderSummarySchema>;

export async function fetchReminders(): Promise<ReminderSummary> {
  return fetchBackendJson("/api/v1/me/reminders", ReminderSummarySchema);
}

// 一覧画面用の単一 "today" authority (R2 F-002)。reference_date (Asia/Tokyo 暦日) + reminder window。
// 一覧 page はこれを一度だけ取得し、全 ticket row の期限強調に同一基準を適用する。
export const DateContextSchema = z.object({
  reference_date: ymdString,
  threshold_days: thresholdDaysSchema
});

export type DateContext = z.infer<typeof DateContextSchema>;

export async function fetchDateContext(): Promise<DateContext> {
  return fetchBackendJson("/api/v1/me/date_context", DateContextSchema);
}
